src/Param.cc

```diff
@@ -428,7 +430,103 @@ bool ParseColorUsingStringStream(const std::string &_input,
 }
 
 //////////////////////////////////////////////////
-bool Param::ValueFromString(const std::string &_value)
+/// \brief Helper function for Param::ValueFromString for parsing pose
+/// This checks the pose components specified in _input are xyzrpy or xyzwxyz
+/// (expects 6 ir 7 values) and the resulting pose is valid.
+/// \param[in] _input Input string.
+/// \param[in] _key Key of the parameter, used for error message.
+/// \param[in] _attributes Attributes associated to this pose, nullptr if there
+/// are none.
+/// \param[out] _value This will be set with the parsed value.
+/// \return True if parsing pose succeeded.
+bool ParsePoseUsingStringStream(const std::string &_input,
+    const std::string &_key, const Param_V *_attributes,
+    ParamPrivate::ParamVariant &_value)
+{
+  StringStreamClassicLocale ss(_input);
+  std::string token;
+  std::vector<double> values;
+  double v;
+  bool isValidPose = true;
+  while (ss >> token)
+  {
+    try
+    {
+      v = std::stod(token);
+    }
+    // Catch invalid argument exception from std::stod
+    catch(std::invalid_argument &)
+    {
+      sdferr << "Invalid argument. Unable to set value ["<< token
+             << "] for key [" << _key << "].\n";
+      isValidPose = false;
+      break;
+    }
+    // Catch out of range exception from std::stof
+    catch(std::out_of_range &)
+    {
+      sdferr << "Out of range. Unable to set value [" << token
+             << "] for key [" << _key << "].\n";
+      isValidPose = false;
+      break;
+    }
+
+    if (values.size() > 7)
+    {
+      sdferr << "Pose values cannot have more than 7 values.\n";
+      isValidPose = false;
+      break;
+    }
+
+    if (!std::isfinite(v))
+    {
+      sdferr << "Pose values must be finite.\n";
+      isValidPose = false;
+      break;
+    }
+
+    values.push_back(v);
+  }
+
+  if (!isValidPose || !_attributes)
+    return ParseUsingStringStream<ignition::math::Pose3d>(_input, _key, _value);
+
+  std::string rotationType = "rpy_radians";
+  for (const auto& p : *_attributes)
+  {
+    if (p->GetKey() == "rotation_type")
+      rotationType = p->GetAsString();
+  }
+
+  if (values.size() == 6u && rotationType == "rpy_radians")
```

Comment:
Error messages regarding size are not concise; consider making errors more specific to the rotation type.

e.g. rather than `if (size == size_for_type && rotationType == type))`, consider using a check which can inject the appropriate error message:
```
auto is_type = [&](type, desired_size) {
  if (rotation_type != type) {
    return false;
  } else if (values.size() != desired_size) {
   sdferr << "The value for //pose[@rotation_type='rpy_radians'] must have " << desired_size << "values, but " << values.size() << " were found instead.";
   return false;
  } else {
   return true;
  }
};
if (is_type("rpy_radians", 6u)) {
  _value = ...
} else if ...
else {
  sdferr << "Invalid type ...";
}
```

Reply:
Thanks for the advice, it's a lot clearer now. Added in https://github.com/ignitionrobotics/sdformat/pull/589/commits/47257c24a652763a511e95572de11c756c12ae1a. Also added more tests to check the error stream.